dali/util/s3_file.cc

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/util/s3_file.h"
+#include <fnmatch.h>
+#include "dali/core/format.h"
+#include "dali/util/s3_client_manager.h"
+#include "dali/util/uri.h"
+namespace dali {
+
+S3FileStream::S3FileStream(Aws::S3::S3Client* s3_client, const std::string& uri,
+                           std::optional<size_t> size)
+    : FileStream(uri), s3_client_(s3_client) {
+  object_location_ = s3_filesystem::parse_uri(uri);
+  if (size.has_value() && size.value() > 0) {
+    object_stats_.exists = true;
+    object_stats_.size = size.value();
+  } else {
+    object_stats_ = s3_filesystem::get_stats(s3_client, object_location_);
+    if (!object_stats_.exists)
+      throw std::runtime_error("S3 Object not found. bucket=" + object_location_.bucket +
+                               " object=" + object_location_.object);
```

Comment:
Related to the other comment, why not just report the error in the get_stats?

Reply:
you mean, just throw it as a proper error and get rid of the exists?